### ISSUE TITLE
Question for Meta-update in Meta-TTS

### ISSUE THREAD

**Issue by v-nhandt21** [SungFeng-Huang/Meta-TTS#4]
I have a question about meta update in the outer loop of Meta TTS

I compare the algorithm of MetaTTS and the original of MAML, are they the same in the below red box

![image](https://user-images.githubusercontent.com/36092323/191220856-8d6ec597-e377-41dd-9789-94c02aa57c79.png)

I can not find the sum of query loss of all tasks in the repo, can you help to show!

![image](https://user-images.githubusercontent.com/36092323/191220630-890da531-c0e7-4654-a900-b4bd0271fd07.png)

I refer to the implementation of meta-learning for mnist, it seems also sum query loss of all tasks:
https://github.com/learnables/learn2learn/blob/0b9d3a3d540646307ca5debf8ad9c79ffe975e1c/examples/vision/meta_mnist.py#L100

One more question is the MAML from L2L, what is the "out of space" mean:
![image](https://user-images.githubusercontent.com/36092323/191224078-472570d2-7eb5-4ee7-826d-e51eca30fe7b.png)

Thank you SungFeng

**Comment by SungFeng-Huang** [comment]
Hi,

The red blocks are actually the same. It is done by pytorch-lightning automatically with distributed training (we use "DDP" in this paper) or accumulate gradients. The `training_step()` function would return the meta-loss of each task, then DDP would synchronize and average the meta-losses of each task on each GPU. The number of tasks to be averaged can be controlled by the number of GPUs and the gradient accumulation steps.

The "out of space" operation is only required for iMAML training, which is implemented but not shown in the paper. The `MAML.adapt()` in L2L does not return a new MAML object, instead it uses the same MAML object (inherited from torch.nn.Module) and simply replace the parameter with the calculated updated tensors. This is called "in-place" in pytorch, and would cause some problems in implementing the iMAML loss. So I inherit my custom MAML from L2L, change the original in-place `adapt()` into `adapt_()` (which is the pytorch coding style of in-place operation), and create an out-of-space `adapt()` (which would return a new MAML object).

**Comment by v-nhandt21** [comment]
@SungFeng-Huang  I am going to reimplement without using pytorch lightning, in the training step, if I have a multispeaker fastspeech with 10 speakers, is that mean I much have 10 cloned models for 10 tasks, can I use randomly 4 tasks at meta_update :))

 I am trying to understand the MAML training process, is the target of MetaTTS is "to train a general model for multispeaker and then fast-tuning for new speaker in the adaptive stage"?

Can the new tuned model work for all speakers (training speaker + new speaker) or only the new speaker?

My target is the new model can work for all speakers?

**Comment by SungFeng-Huang** [comment]
@v-nhandt21 Sure you can randomly choose 4 tasks at meta-update! In my case, my multi-speaker FastSpeech 2 has more than 200 speakers, while each meta-update I use only 8 tasks. If you want to reimplement without pytorch-lightning, since the distributed training is the only way to run tasks in parallel, you might need to handle DDP by yourself, or else you would need gradient accumulation, which is a trade-off between time and computational resources.

Yes, your understanding about MAML/MetaTTS is correct.

In the basic MetaTTS setup, we would fine-tune the whole model for the new speaker, so the tuned model would only work for the new speaker.
You can also get a new tuned model to work for all speakers (training speaker + new speaker) if and only if you extend the speaker embedding lookup table then only tune the newly initialized speaker embedding while keeping the trained speaker embeddings unchanged. But this setting is less efficient, and would not take the "fast-tuning" advantage of MetaTTS as shown in our experiments.

**Comment by v-nhandt21** [comment]
> 

For the speaker encoder, instead of using lookup table, I use d-vecter so that it can work for unseen speakers. I will try this approach and update you soon.

Thank you!
